File: python/liskov-substitution/lsp_violation_demonstration.py
"""
VIOLATION OF LISKOV SUBSTITUTION PRINCIPLE

This demonstrates how LSP can be violated in Python when subclasses
don't properly honor the behavioral contract of their base classes.

Python's dynamic nature makes LSP violations particularly dangerous
because they're often only discovered at runtime.
"""

from abc import ABC, abstractmethod
from typing import List, Optional
import random
from enum import Enum


class DataStructure(ABC):
    """Abstract base class defining the expected contract"""
    
    @abstractmethod
    def add(self, element: int) -> None:
        """Add an element to the data structure"""
        pass
    
    @abstractmethod
    def remove(self) -> int:
        """Remove and return an element"""
        pass
    
    @abstractmethod
    def peek(self) -> int:
        """Look at the next element without removing it"""
        pass
    
    @abstractmethod
    def size(self) -> int:
        """Return the number of elements"""
        pass
    
    @abstractmethod
    def is_empty(self) -> bool:
        """Check if empty"""
        pass


class CorrectStack(DataStructure):
    """Baseline correct implementation"""
    
    def __init__(self):
        self._items: List[int] = []
    
    def add(self, element: int) -> None:
        self._items.append(element)
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Stack is empty")
        return self._items.pop()  # LIFO
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Stack is empty")
        return self._items[-1]
    
    def size(self) -> int:
        return len(self._items)
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"CorrectStack{self._items}"


# VIOLATION 1: Queue masquerading as a Stack
class MislabeledQueue(DataStructure):
    """LSP VIOLATION: Claims to be a stack but behaves like a queue"""
    
    def __init__(self):
        self._items: List[int] = []
    
    def add(self, element: int) -> None:
        self._items.append(element)
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Queue is empty")
        # LSP VIOLATION: FIFO instead of expected LIFO!
        return self._items.pop(0)
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Queue is empty")
        # LSP VIOLATION: Peeks at wrong end!
        return self._items[0]
    
    def size(self) -> int:
        return len(self._items)
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"MislabeledQueue{self._items}"


# VIOLATION 2: Random behavior
class RandomStructure(DataStructure):
    """LSP VIOLATION: Completely unpredictable behavior"""
    
    def __init__(self):
        self._items: List[int] = []
        self._random = random.Random()
    
    def add(self, element: int) -> None:
        self._items.append(element)
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Structure is empty")
        # LSP VIOLATION: Removes random element!
        index = self._random.randint(0, len(self._items) - 1)
        return self._items.pop(index)
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Structure is empty")
        # LSP VIOLATION: Random peek too!
        index = self._random.randint(0, len(self._items) - 1)
        return self._items[index]
    
    def size(self) -> int:
        return len(self._items)
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"RandomStructure{self._items}"


# VIOLATION 3: Behavior changes based on state
class InconsistentStructure(DataStructure):
    """LSP VIOLATION: Behavior changes based on internal state"""
    
    def __init__(self):
        self._items: List[int] = []
    
    def add(self, element: int) -> None:
        self._items.append(element)
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Structure is empty")
        
        # LSP VIOLATION: Behavior depends on size!
        if len(self._items) <= 2:
            return self._items.pop(0)  # FIFO when small
        else:
            return self._items.pop()   # LIFO when large
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Structure is empty")
        
        # LSP VIOLATION: Peek behavior also changes!
        if len(self._items) <= 2:
            return self._items[0]   # Peek front when small
        else:
            return self._items[-1]  # Peek back when large
    
    def size(self) -> int:
        return len(self._items)
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"InconsistentStructure{self._items}"


# VIOLATION 4: Strengthened preconditions
class RestrictiveStructure(DataStructure):
    """LSP VIOLATION: Adds restrictions not in base contract"""
    
    def __init__(self):
        self._items: List[int] = []
    
    def add(self, element: int) -> None:
        # LSP VIOLATION: Strengthened precondition!
        if element < 0:
            raise ValueError("Negative numbers not allowed!")
        if element > 100:
            raise ValueError("Numbers greater than 100 not allowed!")
        self._items.append(element)
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Structure is empty")
        
        # LSP VIOLATION: Additional restriction!
        if len(self._items) == 1:
            raise RuntimeError("Cannot remove last element!")
        
        return self._items.pop()
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Structure is empty")
        
        # LSP VIOLATION: Peek also has restrictions!
        if len(self._items) == 1:
            raise RuntimeError("Cannot peek at last element!")
        
        return self._items[-1]
    
    def size(self) -> int:
        return len(self._items)
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"RestrictiveStructure{self._items}"


# VIOLATION 5: Hidden side effects
class SideEffectStructure(DataStructure):
    """LSP VIOLATION: Operations have unexpected side effects"""
    
    def __init__(self):
        self._items: List[int] = []
        self._operation_count = 0
        self._hidden_log: List[str] = []
    
    def add(self, element: int) -> None:
        self._items.append(element)
        self._operation_count += 1
        
        # LSP VIOLATION: Hidden side effect!
        if self._operation_count % 3 == 0:
            self._hidden_log.append(f"Secret: Added {element}")
            # Secretly double the element!
            self._items.append(element)
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Structure is empty")
        
        element = self._items.pop()
        self._operation_count += 1
        
        # LSP VIOLATION: Hidden side effect!
        if self._operation_count % 5 == 0:
            self._hidden_log.append(f"Secret: Removed {element}")
            # Secretly remove another element!
            if self._items:
                self._items.pop()
        
        return element
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Structure is empty")
        
        # LSP VIOLATION: Even peek has side effects!
        self._operation_count += 1
        self._hidden_log.append(f"Peeked at {self._items[-1]}")
        
        return self._items[-1]
    
    def size(self) -> int:
        # LSP VIOLATION: Size operation has side effects!
        self._operation_count += 1
        return len(self._items)
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"SideEffectStructure{self._items}"
    
    def get_hidden_log(self) -> List[str]:
        """This method reveals the hidden operations"""
        return self._hidden_log.copy()


# VIOLATION 6: Weakened postconditions  
class WeakStructure(DataStructure):
    """LSP VIOLATION: Doesn't guarantee what base class promises"""
    
    def __init__(self):
        self._items: List[int] = []
        self._corrupt_chance = 0.1
    
    def add(self, element: int) -> None:
        self._items.append(element)
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Structure is empty")
        
        # LSP VIOLATION: Sometimes returns wrong value!
        if random.random() < self._corrupt_chance:
            return -999  # Corrupt value instead of actual element
        
        return self._items.pop()
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Structure is empty")
        
        # LSP VIOLATION: Sometimes returns wrong value!
        if random.random() < self._corrupt_chance:
            return -999  # Corrupt value
        
        return self._items[-1]
    
    def size(self) -> int:
        # LSP VIOLATION: Sometimes returns wrong size!
        real_size = len(self._items)
        if random.random() < self._corrupt_chance:
            return real_size + 1  # Wrong size
        return real_size
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"WeakStructure{self._items}"


# Client function that expects consistent behavior
def process_data_structure(ds: DataStructure, name: str) -> None:
    """Process any data structure - expects consistent behavior"""
    print(f"--- Processing {name} ---")
    
    try:
        # Add elements
        ds.add(10)
        ds.add(20)
        ds.add(30)
        print(f"Added 10, 20, 30. Size: {ds.size()}")
        
        # Peek and remove
        peeked = ds.peek()
        print(f"Peeked: {peeked}")
        
        removed = ds.remove()
        print(f"Removed: {removed}")
        
        # Expect peek and remove to match for stack behavior
        if peeked != removed:
            print(f"⚠️  WARNING: Peek ({peeked}) != Remove ({removed})!")
        
        print(f"Size after removal: {ds.size()}")
        print(f"Is empty: {ds.is_empty()}")
        
    except Exception as e:
        print(f"ERROR: {e}")
    
    print()


def reverse_data_structure(ds: DataStructure) -> None:
    """Tries to reverse elements - works correctly only with proper stacks"""
    print(f"Attempting to reverse: {ds}")
    
    try:
        # Remove all elements
        temp = []
        while not ds.is_empty():
            temp.append(ds.remove())
        
        # Add them back (should reverse for stack)
        for element in temp:
            ds.add(element)
        
        print(f"After reverse attempt: {ds}")
    
    except Exception as e:
        print(f"Reverse failed: {e}")


def demonstrate_violations():
    """Show how different violations break client expectations"""
    print("=== TESTING WITH PROBLEMATIC IMPLEMENTATIONS ===")
    
    structures = [
        CorrectStack(),
        MislabeledQueue(),
        RandomStructure(),
        InconsistentStructure(),
        RestrictiveStructure(),
        SideEffectStructure(),
        WeakStructure()
    ]
    
    names = [
        "CorrectStack (Baseline)",
        "MislabeledQueue (FIFO Violation)",
        "RandomStructure (Unpredictable)",
        "InconsistentStructure (State-dependent)",
        "RestrictiveStructure (Precondition Violation)",
        "SideEffectStructure (Hidden Effects)",
        "WeakStructure (Postcondition Violation)"
    ]
    
    for ds, name in zip(structures, names):
        process_data_structure(ds, name)


def demonstrate_broken_client_code():
    """Show how LSP violations break client code"""
    print("=== DEMONSTRATING BROKEN CLIENT CODE ===")
    
    test_structures = [
        (CorrectStack(), "CorrectStack (works)"),
        (MislabeledQueue(), "MislabeledQueue (broken)"),
        (InconsistentStructure(), "InconsistentStructure (unpredictable)")
    ]
    
    for ds, name in test_structures:
        # Add test elements
        ds.add(1)
        ds.add(2)
        ds.add(3)
        
        print(f"\nReverse test with {name}:")
        reverse_data_structure(ds)


def demonstrate_precondition_violations():
    """Show how strengthened preconditions break LSP"""
    print("\n=== PRECONDITION VIOLATION DEMONSTRATION ===")
    
    restrictive = RestrictiveStructure()
    
    print("Testing RestrictiveStructure:")
    try:
        restrictive.add(5)
        restrictive.add(10)
        print("Added 5, 10 successfully")
        
        print("Trying to add -5 (should fail):")
        restrictive.add(-5)
    except Exception as e:
        print(f"Failed as expected: {e}")
    
    try:
        print("Trying to add 150 (should fail):")
        restrictive.add(150)
    except Exception as e:
        print(f"Failed as expected: {e}")


def demonstrate_side_effects():
    """Show how side effects violate LSP"""
    print("\n=== SIDE EFFECT DEMONSTRATION ===")
    
    side_effect_ds = SideEffectStructure()
    
    print("Testing SideEffectStructure:")
    print("Adding elements 1, 2, 3, 4, 5:")
    
    for i in range(1, 6):
        side_effect_ds.add(i)
        print(f"After adding {i}: size = {side_effect_ds.size()}")
    
    print(f"Final structure: {side_effect_ds}")
    print(f"Hidden log: {side_effect_ds.get_hidden_log()}")


def main():
    print("=== LSP VIOLATION DEMONSTRATION ===")
    print("Objects implementing same interface but violating behavioral contracts")
    print()
    
    demonstrate_violations()
    demonstrate_broken_client_code()
    demonstrate_precondition_violations()
    demonstrate_side_effects()
    
    print("\n=== WHY THESE VIOLATE LSP ===")
    print("1. MislabeledQueue: Changes expected removal order (FIFO vs LIFO)")
    print("2. RandomStructure: Unpredictable behavior breaks client expectations")
    print("3. InconsistentStructure: Behavior changes based on internal state")
    print("4. RestrictiveStructure: Strengthens preconditions with new restrictions")
    print("5. SideEffectStructure: Hidden side effects violate behavioral contract")
    print("6. WeakStructure: Weakens postconditions by returning wrong values")
    print("7. Client code written for base class fails with these implementations")
    print("8. Substitutability is broken - cannot replace base with derived safely")


if __name__ == "__main__":
    main()

File: rust/liskov-substitution/lsp_violation_demonstration.rs
// VIOLATION OF LISKOV SUBSTITUTION PRINCIPLE
//
// This demonstrates how LSP can be violated in Rust when types
// that implement the same trait behave in ways that break
// client expectations.
//
// Rust's type system helps prevent many LSP violations at compile time,
// but behavioral violations can still occur at runtime.

use std::fmt::Debug;
use rand::Rng;

/// Trait defining the expected contract
pub trait DataStructure: Debug {
    /// Add an element to the data structure
    fn add(&mut self, element: i32) -> Result<(), String>;
    
    /// Remove and return an element
    fn remove(&mut self) -> Result<i32, String>;
    
    /// Look at the next element without removing it
    fn peek(&self) -> Result<i32, String>;
    
    /// Return the number of elements
    fn size(&self) -> usize;
    
    /// Check if empty
    fn is_empty(&self) -> bool {
        self.size() == 0
    }
}

/// Correct baseline implementation
#[derive(Debug, Clone)]
pub struct CorrectStack {
    items: Vec<i32>,
}

impl CorrectStack {
    pub fn new() -> Self {
        CorrectStack {
            items: Vec::new(),
        }
    }
}

impl DataStructure for CorrectStack {
    fn add(&mut self, element: i32) -> Result<(), String> {
        self.items.push(element);
        Ok(())
    }
    
    fn remove(&mut self) -> Result<i32, String> {
        self.items.pop().ok_or_else(|| "Stack is empty".to_string())
    }
    
    fn peek(&self) -> Result<i32, String> {
        self.items.last().copied().ok_or_else(|| "Stack is empty".to_string())
    }
    
    fn size(&self) -> usize {
        self.items.len()
    }
}

// VIOLATION 1: Queue masquerading as a Stack
#[derive(Debug, Clone)]
pub struct MislabeledQueue {
    items: Vec<i32>,
}

impl MislabeledQueue {
    pub fn new() -> Self {
        MislabeledQueue {
            items: Vec::new(),
        }
    }
}

impl DataStructure for MislabeledQueue {
    fn add(&mut self, element: i32) -> Result<(), String> {
        self.items.push(element);
        Ok(())
    }
    
    // LSP VIOLATION: FIFO instead of expected LIFO!
    fn remove(&mut self) -> Result<i32, String> {
        if self.items.is_empty() {
            Err("Queue is empty".to_string())
        } else {
            Ok(self.items.remove(0)) // FIFO behavior
        }
    }
    
    // LSP VIOLATION: Peeks at wrong end!
    fn peek(&self) -> Result<i32, String> {
        self.items.first().copied().ok_or_else(|| "Queue is empty".to_string())
    }
    
    fn size(&self) -> usize {
        self.items.len()
    }
}

// VIOLATION 2: Random behavior structure
#[derive(Debug)]
pub struct RandomStructure {
    items: Vec<i32>,
    rng: rand::rngs::ThreadRng,
}

impl RandomStructure {
    pub fn new() -> Self {
        RandomStructure {
            items: Vec::new(),
            rng: rand::thread_rng(),
        }
    }
}

impl DataStructure for RandomStructure {
    fn add(&mut self, element: i32) -> Result<(), String> {
        self.items.push(element);
        Ok(())
    }
    
    // LSP VIOLATION: Removes random element!
    fn remove(&mut self) -> Result<i32, String> {
        if self.items.is_empty() {
            return Err("Structure is empty".to_string());
        }
        
        let index = self.rng.gen_range(0..self.items.len());
        Ok(self.items.remove(index))
    }
    
    // LSP VIOLATION: Random peek too!
    fn peek(&self) -> Result<i32, String> {
        if self.items.is_empty() {
            return Err("Structure is empty".to_string());
        }
        
        // Since we can't mutate rng in peek, we'll use a different approach
        let index = self.items.len() % (self.items.len().max(1));
        Ok(self.items[index])
    }
    
    fn size(&self) -> usize {
        self.items.len()
    }
}

// VIOLATION 3: Behavior changes based on state
#[derive(Debug, Clone)]
pub struct InconsistentStructure {
    items: Vec<i32>,
}

impl InconsistentStructure {
    pub fn new() -> Self {
        InconsistentStructure {
            items: Vec::new(),
        }
    }
}

impl DataStructure for InconsistentStructure {
    fn add(&mut self, element: i32) -> Result<(), String> {
        self.items.push(element);
        Ok(())
    }
    
    // LSP VIOLATION: Behavior depends on size!
    fn remove(&mut self) -> Result<i32, String> {
        if self.items.is_empty() {
            return Err("Structure is empty".to_string());
        }
        
        if self.items.len() <= 2 {
            Ok(self.items.remove(0)) // FIFO when small
        } else {
            self.items.pop().ok_or_else(|| "Structure is empty".to_string()) // LIFO when large
        }
    }
    
    // LSP VIOLATION: Peek behavior also changes!
    fn peek(&self) -> Result<i32, String> {
        if self.items.is_empty() {
            return Err("Structure is empty".to_string());
        }
        
        if self.items.len() <= 2 {
            Ok(self.items[0]) // Peek front when small
        } else {
            Ok(self.items[self.items.len() - 1]) // Peek back when large
        }
    }
    
    fn size(&self) -> usize {
        self.items.len()
    }
}

// VIOLATION 4: Strengthened preconditions
#[derive(Debug, Clone)]
pub struct RestrictiveStructure {
    items: Vec<i32>,
}

impl RestrictiveStructure {
    pub fn new() -> Self {
        RestrictiveStructure {
            items: Vec::new(),
        }
    }
}

impl DataStructure for RestrictiveStructure {
    // LSP VIOLATION: Strengthened precondition!
    fn add(&mut self, element: i32) -> Result<(), String> {
        if element < 0 {
            return Err("Negative numbers not allowed!".to_string());
        }
        if element > 100 {
            return Err("Numbers greater than 100 not allowed!".to_string());
        }
        self.items.push(element);
        Ok(())
    }
    
    // LSP VIOLATION: Additional restriction!
    fn remove(&mut self) -> Result<i32, String> {
        if self.items.is_empty() {
            return Err("Structure is empty".to_string());
        }
        
        if self.items.len() == 1 {
            return Err("Cannot remove last element!".to_string());
        }
        
        self.items.pop().ok_or_else(|| "Structure is empty".to_string())
    }
    
    // LSP VIOLATION: Peek also has restrictions!
    fn peek(&self) -> Result<i32, String> {
        if self.items.is_empty() {
            return Err("Structure is empty".to_string());
        }
        
        if self.items.len() == 1 {
            return Err("Cannot peek at last element!".to_string());
        }
        
        Ok(self.items[self.items.len() - 1])
    }
    
    fn size(&self) -> usize {
        self.items.len()
    }
}

// VIOLATION 5: Hidden side effects
#[derive(Debug, Clone)]
pub struct SideEffectStructure {
    items: Vec<i32>,
    operation_count: usize,
    hidden_log: Vec<String>,
}

impl SideEffectStructure {
    pub fn new() -> Self {
        SideEffectStructure {
            items: Vec::new(),
            operation_count: 0,
            hidden_log: Vec::new(),
        }
    }
    
    pub fn get_hidden_log(&self) -> &[String] {
        &self.hidden_log
    }
}

impl DataStructure for SideEffectStructure {
    // LSP VIOLATION: Hidden side effects!
    fn add(&mut self, element: i32) -> Result<(), String> {
        self.items.push(element);
        self.operation_count += 1;
        
        // Hidden side effect!
        if self.operation_count % 3 == 0 {
            self.hidden_log.push(format!("Secret: Added {}", element));
            // Secretly double the element!
            self.items.push(element);
        }
        
        Ok(())
    }
    
    // LSP VIOLATION: Hidden side effects in remove!
    fn remove(&mut self) -> Result<i32, String> {
        if self.items.is_empty() {
            return Err("Structure is empty".to_string());
        }
        
        let element = self.items.pop().unwrap();
        self.operation_count += 1;
        
        // Hidden side effect!
        if self.operation_count % 5 == 0 {
            self.hidden_log.push(format!("Secret: Removed {}", element));
            // Secretly remove another element!
            if !self.items.is_empty() {
                self.items.pop();
            }
        }
        
        Ok(element)
    }
    
    // LSP VIOLATION: Even peek has side effects!
    fn peek(&self) -> Result<i32, String> {
        if self.items.is_empty() {
            return Err("Structure is empty".to_string());
        }
        
        // We can't modify self in peek, but this shows the concept
        // In a real implementation, this might use interior mutability
        
        self.items.last().copied().ok_or_else(|| "Structure is empty".to_string())
    }
    
    fn size(&self) -> usize {
        self.items.len()
    }
}

// VIOLATION 6: Weakened postconditions
#[derive(Debug, Clone)]
pub struct WeakStructure {
    items: Vec<i32>,
}

impl WeakStructure {
    pub fn new() -> Self {
        WeakStructure {
            items: Vec::new(),
        }
    }
}

impl DataStructure for WeakStructure {
    fn add(&mut self, element: i32) -> Result<(), String> {
        self.items.push(element);
        Ok(())
    }
    
    // LSP VIOLATION: Sometimes returns wrong value!
    fn remove(&mut self) -> Result<i32, String> {
        if self.items.is_empty() {
            return Err("Structure is empty".to_string());
        }
        
        // Simulate corruption
        if rand::random::<f64>() < 0.1 {
            self.items.pop(); // Remove the element but return wrong value
            Ok(-999) // Corrupt value
        } else {
            self.items.pop().ok_or_else(|| "Structure is empty".to_string())
        }
    }
    
    // LSP VIOLATION: Sometimes returns wrong value!
    fn peek(&self) -> Result<i32, String> {
        if self.items.is_empty() {
            return Err("Structure is empty".to_string());
        }
        
        // Simulate corruption in peek too
        if rand::random::<f64>() < 0.1 {
            Ok(-999) // Corrupt value
        } else {
            Ok(self.items[self.items.len() - 1])
        }
    }
    
    // LSP VIOLATION: Sometimes returns wrong size!
    fn size(&self) -> usize {
        let real_size = self.items.len();
        if rand::random::<f64>() < 0.1 {
            real_size.saturating_add(1) // Wrong size
        } else {
            real_size
        }
    }
}

/// Client function that expects consistent behavior
pub fn process_data_structure(ds: &mut dyn DataStructure, name: &str) {
    println!("--- Processing {} ---", name);
    
    // Add elements
    match ds.add(10) {
        Ok(()) => {},
        Err(e) => {
            println!("Add error: {}", e);
            return;
        }
    }
    match ds.add(20) {
        Ok(()) => {},
        Err(e) => {
            println!("Add error: {}", e);
            return;
        }
    }
    match ds.add(30) {
        Ok(()) => {},
        Err(e) => {
            println!("Add error: {}", e);
            return;
        }
    }
    
    println!("Added 10, 20, 30. Size: {}", ds.size());
    
    // Peek and remove
    let peeked = ds.peek();
    let removed = ds.remove();
    
    match (peeked, removed) {
        (Ok(peek_val), Ok(remove_val)) => {
            println!("Peeked: {}, Removed: {}", peek_val, remove_val);
            if peek_val != remove_val {
                println!("⚠️  WARNING: Peek ({}) != Remove ({})!", peek_val, remove_val);
            }
        }
        (Ok(peek_val), Err(e)) => println!("Peeked: {}, Remove error: {}", peek_val, e),
        (Err(e), Ok(remove_val)) => println!("Peek error: {}, Removed: {}", e, remove_val),
        (Err(e1), Err(e2)) => println!("Peek error: {}, Remove error: {}", e1, e2),
    }
    
    println!("Size after removal: {}", ds.size());
    println!("Is empty: {}", ds.is_empty());
    println!();
}

/// Tries to reverse elements - works correctly only with proper stacks
pub fn reverse_data_structure(ds: &mut dyn DataStructure) {
    println!("Attempting to reverse: {:?}", ds);
    
    // Remove all elements
    let mut temp = Vec::new();
    while !ds.is_empty() {
        match ds.remove() {
            Ok(element) => temp.push(element),
            Err(_) => break,
        }
    }
    
    // Add them back (should reverse for stack)
    for element in temp {
        if ds.add(element).is_err() {
            println!("Failed to add element back during reverse");
            break;
        }
    }
    
    println!("After reverse attempt: {:?}", ds);
}

fn demonstrate_violations() {
    println!("=== TESTING WITH PROBLEMATIC IMPLEMENTATIONS ===");
    
    let mut structures: Vec<Box<dyn DataStructure>> = vec![
        Box::new(CorrectStack::new()),
        Box::new(MislabeledQueue::new()),
        Box::new(RandomStructure::new()),
        Box::new(InconsistentStructure::new()),
        Box::new(RestrictiveStructure::new()),
        Box::new(SideEffectStructure::new()),
        Box::new(WeakStructure::new()),
    ];
    
    let names = [
        "CorrectStack (Baseline)",
        "MislabeledQueue (FIFO Violation)",
        "RandomStructure (Unpredictable)",
        "InconsistentStructure (State-dependent)",
        "RestrictiveStructure (Precondition Violation)",
        "SideEffectStructure (Hidden Effects)",
        "WeakStructure (Postcondition Violation)",
    ];
    
    for (ds, name) in structures.iter_mut().zip(names.iter()) {
        process_data_structure(ds.as_mut(), name);
    }
}

fn demonstrate_broken_client_code() {
    println!("=== DEMONSTRATING BROKEN CLIENT CODE ===");
    
    let mut test_structures: Vec<(Box<dyn DataStructure>, &str)> = vec![
        (Box::new(CorrectStack::new()), "CorrectStack (works)"),
        (Box::new(MislabeledQueue::new()), "MislabeledQueue (broken)"),
        (Box::new(InconsistentStructure::new()), "InconsistentStructure (unpredictable)"),
    ];
    
    for (ds, name) in test_structures.iter_mut() {
        // Add test elements
        let _ = ds.add(1);
        let _ = ds.add(2);
        let _ = ds.add(3);
        
        println!("\nReverse test with {}:", name);
        reverse_data_structure(ds.as_mut());
    }
}

fn demonstrate_precondition_violations() {
    println!("\n=== PRECONDITION VIOLATION DEMONSTRATION ===");
    
    let mut restrictive = RestrictiveStructure::new();
    
    println!("Testing RestrictiveStructure:");
    
    match restrictive.add(5) {
        Ok(()) => println!("Added 5 successfully"),
        Err(e) => println!("Failed to add 5: {}", e),
    }
    
    match restrictive.add(10) {
        Ok(()) => println!("Added 10 successfully"),
        Err(e) => println!("Failed to add 10: {}", e),
    }
    
    println!("Trying to add -5 (should fail):");
    match restrictive.add(-5) {
        Ok(()) => println!("⚠️  Unexpectedly succeeded!"),
        Err(e) => println!("Failed as expected: {}", e),
    }
    
    println!("Trying to add 150 (should fail):");
    match restrictive.add(150) {
        Ok(()) => println!("⚠️  Unexpectedly succeeded!"),
        Err(e) => println!("Failed as expected: {}", e),
    }
}

fn demonstrate_side_effects() {
    println!("\n=== SIDE EFFECT DEMONSTRATION ===");
    
    let mut side_effect_ds = SideEffectStructure::new();
    
    println!("Testing SideEffectStructure:");
    println!("Adding elements 1, 2, 3, 4, 5:");
    
    for i in 1..=5 {
        match side_effect_ds.add(i) {
            Ok(()) => println!("After adding {}: size = {}", i, side_effect_ds.size()),
            Err(e) => println!("Failed to add {}: {}", i, e),
        }
    }
    
    println!("Final structure: {:?}", side_effect_ds);
    println!("Hidden log: {:?}", side_effect_ds.get_hidden_log());
}

pub fn main() {
    println!("=== LSP VIOLATION DEMONSTRATION ===");
    println!("Objects implementing same trait but violating behavioral contracts");
    println!();
    
    demonstrate_violations();
    demonstrate_broken_client_code();
    demonstrate_precondition_violations();
    demonstrate_side_effects();
    
    println!("\n=== WHY THESE VIOLATE LSP ===");
    println!("1. MislabeledQueue: Changes expected removal order (FIFO vs LIFO)");
    println!("2. RandomStructure: Unpredictable behavior breaks client expectations");
    println!("3. InconsistentStructure: Behavior changes based on internal state");
    println!("4. RestrictiveStructure: Strengthens preconditions with new restrictions");
    println!("5. SideEffectStructure: Hidden side effects violate behavioral contract");
    println!("6. WeakStructure: Weakens postconditions by returning wrong values");
    println!("7. Client code written for base trait fails with these implementations");
    println!("8. Substitutability is broken - cannot replace base with derived safely");
}